codec-http2/src/main/java/io/netty/handler/codec/http2/UniformStreamByteDistributor.java

```diff
@@ -0,0 +1,203 @@
+import static java.lang.Math.max;
+import static java.lang.Math.min;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+
+/**
+ * A {@link StreamByteDistributor} that ignores stream priority and uniformly allocates bytes to all
+ * streams. This class uses a minimum chunk size that will be allocated to each stream. While
+ * fewer streams may be written to in each call to {@link #distribute(int, Writer)}, doing this
+ * should improve the goodput on each written stream.
+ */
+public final class UniformStreamByteDistributor implements StreamByteDistributor {
+    static final int DEFAULT_MIN_ALLOCATION_CHUNK = 1024;
+
+    private final Http2Connection.PropertyKey stateKey;
+    private final Deque<State> queue = new ArrayDeque<State>();
+    private final Deque<State> emptyFrameQueue = new ArrayDeque<State>();
+
+    /**
+     * The minimum number of bytes that we will attempt to allocate to a stream. This is to
+     * help improve goodput on a per-stream basis.
+     */
+    private int minAllocationChunk = DEFAULT_MIN_ALLOCATION_CHUNK;
+    private long totalStreamableBytes;
+
+    public UniformStreamByteDistributor(Http2Connection connection) {
+        checkNotNull(connection, "connection");
+
+        // Add a state for the connection.
+        stateKey = connection.newKey();
+        Http2Stream connectionStream = connection.connectionStream();
+        connectionStream.setProperty(stateKey, new State(connectionStream));
+
+        // Register for notification of new streams.
+        connection.addListener(new Http2ConnectionAdapter() {
+            @Override
+            public void onStreamAdded(Http2Stream stream) {
+                stream.setProperty(stateKey, new State(stream));
+            }
+
+            @Override
+            public void onStreamClosed(Http2Stream stream) {
+                state(stream).close();
+            }
+        });
+    }
+
+    /**
+     * Sets the minimum allocation chunk that will be allocated to each stream. Defaults to 1KiB.
+     *
+     * @param minAllocationChunk the minimum number of bytes that will be allocated to each stream.
+     * Must be > 0.
+     */
+    public void minAllocationChunk(int minAllocationChunk) {
+        if (minAllocationChunk <= 0) {
+            throw new IllegalArgumentException("minAllocationChunk must be > 0");
+        }
+        this.minAllocationChunk = minAllocationChunk;
+    }
+
+    @Override
+    public void updateStreamableBytes(StreamState streamState) {
+        State state = state(streamState.stream());
+        state.updateStreamableBytes(streamState.streamableBytes(), streamState.hasFrame());
+    }
+
+    @Override
+    public boolean distribute(int maxBytes, Writer writer) throws Http2Exception {
+        checkNotNull(writer, "writer");
+
+        // First, write out any empty frames.
+        while (!emptyFrameQueue.isEmpty()) {
+            State state = emptyFrameQueue.remove();
```

Comment:
I'm in favor of keeping 1 queue and sending empty frames immediately (if possible, i.e. nothing is already queued).  Keeping 2 queues seems to be more complicated and it can be confusing that state in the `queue` may also have empty frames to send. These 2 different queues have different ordering semantics to conceptualize and coordinate.  For example what happens if a `State` object initially has an empty frame, and then a non-empty frame before a `write` event takes place? It seems like we use `enqueued` to indicate the `State` object is in `emptyFrameQueue` or `queue`, and then enforce that we can only add the `State` object to 1 of the queues (not both).